Time Validity PR

rgottsc opened the pull request:
Time validity of a dataset is now displayed

rgottsc commented:
#83

ChristianBeilschmidt commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Time validity indicator in data repository](https://trello.com/c/DcDEuweR/1-time-validity-indicator-in-data-repository)

ChristianBeilschmidt commented on projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts:
```diff
@@ -102,6 +109,8 @@ export interface MappingSourceDict {
     channels?: [MappingSourceRasterLayerDict];
     layer?: [MappingSourceVectorLayerDict];
+    time_start?: Time;
+    time_end?: Time;
```
The source delivers strings and not `Time` types I guess. This is why you don't construct these types but just assign them. But in the end, `toString` of `Time` is not called. You see this in the different output format.

rgottsc replied:
fixed in 9919298afbd39ba8285c6946f1eec93e1c65b5ba

ChristianBeilschmidt commented on projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts:
```diff
@@ -131,6 +140,8 @@ export interface MappingSourceRasterLayerDict {
         citation: string,
     };
+    time_start?: Time;
+    time_end?: Time;
```
Same here. `…Dict`s always reflect the serialized forms.

rgottsc replied:
fixed in 9919298afbd39ba8285c6946f1eec93e1c65b5ba

ChristianBeilschmidt commented on projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts:
```diff
@@ -154,6 +165,8 @@ export interface MappingSourceVectorLayerDict {
         citation: string,
     };
+    time_start?: Time;
+    time_end?: Time;
```
Same here.

rgottsc replied:
fixed in 9919298afbd39ba8285c6946f1eec93e1c65b5ba

ChristianBeilschmidt commented on projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts:
```diff
@@ -34,7 +34,7 @@ export class SourceDatasetComponent implements OnInit {
     _showDescription = false;
     _channelSource: ChannelDataSource;
-    _displayedColumns = ['name', 'measurement'];
+    _displayedColumns = ['name', 'measurement', 'start', 'end'];
```
You should display `start` and `end` only if there is information available.

rgottsc replied:
fixed in bc0236c100809397960ca46b0f1e76b973d72473

ChristianBeilschmidt commented on projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts:
```diff
@@ -33,7 +33,7 @@ export class VectorSourceDatasetComponent implements OnInit {
     _showDescription = false;
     _tableSource: LayerTableDataSource;
-    _displayedColumns  = ['title'];
+    _displayedColumns  = ['title', 'start', 'end'];
```
Same here.

rgottsc replied:
fixed in bc0236c100809397960ca46b0f1e76b973d72473

ChristianBeilschmidt commented on projects/wave-core/src/lib/users/user.service.ts:
```diff
@@ -477,6 +480,8 @@ export class UserService {
             textual: layer.textual || [],
             numeric: layer.numeric || [],
+            time_start: !!layer.time_start ? new TimePoint(layer.time_start) : new TimePoint('') ,
+            time_end: !!layer.time_end ? new TimePoint(layer.time_end) : new TimePoint('') ,
```
Both fields are marked as optional on the type `SourceVectorLayerDescription`, so it could just be undefined if the field is not present.

rgottsc replied:
fixed in 6da1d4c74b7a75a702074ebf5ef83c4bda092d4f

ChristianBeilschmidt commented on projects/wave-core/src/lib/users/user.service.ts:
```diff
@@ -516,6 +521,8 @@ export class UserService {
             coords: coords as { crs: string, origin: number[], scale: number[], size: number[] },
             provenance: channelProvenance,
+            time_start: !!channel.time_start ? new TimePoint(channel.time_start) : new TimePoint('') ,
+            time_end: !!channel.time_end ? new TimePoint(channel.time_end) : new TimePoint('') ,
```
Same here.

rgottsc replied:
fixed in 6da1d4c74b7a75a702074ebf5ef83c4bda092d4f

ChristianBeilschmidt commented on projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts:
```diff
@@ -25,7 +25,7 @@ import {NotificationService} from '../../../../notification.service';
 })
 
-export class VectorSourceDatasetComponent implements OnInit {
+export class VectorSourceDatasetComponent implements OnInit , OnChanges  {
```
Please format the code.

rgottsc replied:
formated in 6da1d4c74b7a75a702074ebf5ef83c4bda092d4f****

ChristianBeilschmidt commented on projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts:
```diff
@@ -26,15 +26,15 @@ import {Colormap} from '../../../../colors/colormaps/colormap.model';
 })
 
-export class SourceDatasetComponent implements OnInit {
+export class SourceDatasetComponent implements OnInit , OnChanges {
```
Same here.

rgottsc replied:
formated in 6da1d4c74b7a75a702074ebf5ef83c4bda092d4f

ChristianBeilschmidt commented on projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts:
```diff
@@ -26,15 +26,15 @@ import {Colormap} from '../../../../colors/colormaps/colormap.model';
     _showDescription = false;
     _channelSource: ChannelDataSource;
-    _displayedColumns = ['name', 'measurement'];
+    _displayedColumns = ['name', 'measurement'] ;
```
Please revert this change.

rgottsc replied:
whitespaced deleted in 6da1d4c74b7a75a702074ebf5ef83c4bda092d4f

ChristianBeilschmidt commented on projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts:
```diff
@@ -96,6 +96,68 @@ export class VectorSourceDatasetComponent implements OnInit {
+                        this.dataset.vectorLayer.forEach((element) => {
+                            for (const sourceVar in element) {
+                                if (element.hasOwnProperty(sourceVar)) {
+                                    if (sourceVar === 'time_start') {
```
You know the type of `element`, so you can access the fields directly without needing to iterate over them, check field names, etc.

ChristianBeilschmidt replied:
Its the same for the checks in source-dataset.component.ts

rgottsc replied:
deleted the useless for-loop in 6da1d4c74b7a75a702074ebf5ef83c4bda092d4f

ChristianBeilschmidt commented on projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts:
```diff
@@ -96,6 +96,68 @@ export class VectorSourceDatasetComponent implements OnInit {
+                        });
+                        // Update the global Min/Max in the whole dataset
+                        this.dataset.time_start = globalDatasetMin;
+                        this.dataset.time_end = globalDatasetMax;
```
If you change the dataset, then move this processing to the `UserService`.

rgottsc replied:
reverted the dataset manipulation/processing part in bcf99b0df6ed19d67ce802b25df192cf7edd124e

As discussed i would like to do the dataset processing part in a new PR

ChristianBeilschmidt commented on projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts:
```diff
@@ -96,6 +96,68 @@ export class VectorSourceDatasetComponent implements OnInit {
+                                if (element.hasOwnProperty(sourceVar)) {
+                                    if (sourceVar === 'time_start') {
+                                        if (!element.time_start.isValid()) {
+                                            element.time_start = globalDatasetMin;
```
This is unfortunately incorrect. The rule is: If the layer has no own time interval, use the one from the dataset.
If you changed the dataset's validity before to display the global min/max, then this potentially produces wrong intervals.

rgottsc replied:
i hope its fixed in 6da1d4c74b7a75a702074ebf5ef83c4bda092d4f

rgottsc replied:
reverted the dataset manipulation/processing part in bcf99b0df6ed19d67ce802b25df192cf7edd124e

As discussed i would like to do the dataset processing part in a new PR

ChristianBeilschmidt reviewed (changes requested):
I have some comments regarding the computations. Furthermore, please address my issues one by one (also the ones before). Otherwise, I can't resolve them to see if I can merge the PR.

ChristianBeilschmidt commented on projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.html:
```diff
@@ -16,12 +16,16 @@
 
     <mat-list>
+        <mat-list-item *ngIf="dataset?.time_start || dataset?.time_end">
+            <p><b>Time validity:</b> {{dataset?.time_start.toString()}} - {{dataset?.time_end.toString()}}</p>
```
```suggestion
            <p><b>Time validity:</b> {{dataset?.time_start}} - {{dataset?.time_end}}</p>
```
I think `toString()` is called automatically in templates, so it is redundant and you do it differently below.

rgottsc replied:
fixed in 900ce5db69b7712d277d6400055ede30c92368c1

ChristianBeilschmidt commented on projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.html:
```diff
@@ -12,6 +12,9 @@
 
     <mat-list>
+        <mat-list-item *ngIf="dataset?.time_start || dataset?.time_end">
+            <p><b>Time validity:</b> {{dataset?.time_start.toString()}} - {{dataset?.time_end.toString()}}</p>
```
```suggestion
            <p><b>Time validity:</b> {{dataset?.time_start}} - {{dataset?.time_end}}</p>
```

same here

rgottsc replied:
fixed in 900ce5db69b7712d277d6400055ede30c92368c1

ChristianBeilschmidt commented on projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts:
```diff
@@ -96,6 +96,29 @@ export class VectorSourceDatasetComponent implements OnInit {
+                    // check if there is any time-validity start/end data. If there is start/end data show the column of this data.
+                    case 'dataset': {
+                        this.dataset.rasterLayer.forEach((element) => {
+                            if (element.time_start.isValid()) {
```
What happens if `time_start` is undefined?

rgottsc replied:
fixed in 1a0b23afc2302b363ceb11f273d325f130d8539d

ChristianBeilschmidt reviewed (changes requested):
There are two minor things and one potential bug. The rest looks fine.

ChristianBeilschmidt commented:
![grafik](https://user-images.githubusercontent.com/4947061/105707424-34377880-5f13-11eb-9cee-0a68c84c830a.png)

Can you look at these datasets and figure out, why the wrong time validity (now) is shown?

rgottsc commented:
Unfortunately, I did not implement the undefined case correctly at one point in `user-service.ts`. It is now fixed. Furthermore I changed the ng-if in `source-dataset.component.htm`l and `vector-source-dataset.comonent.html`  a little bit, so that the time validity is only shown if start and end are not undefined
